### ISSUE TITLE
How to format incoming requests w multiple images?

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer-app#6]
Should it be a Json format with an array of steps, and should we start by importing a list of URLs and assigning them short variable names in the Json? 

Or should we basically be submitting a Js function that runs using an array of URLs as additional input?

**Comment by jywarren** [comment]
@tech4gt how did you address this in the initial implementation? Thanks!!

**Comment by tech4GT** [comment]
Hi Jeff, right now I am taking the urls as an array in the request json and the steps are taken as a sequencer string.

**Comment by jywarren** [comment]
Oh ok and how do we map from the array to the steps? I guess we can just do
it manually for now and figure out a flexible architecture later. Thanks!

**Comment by harshkhandeparkar** [comment]
Is there a way to compress the images if they exceed a limit. Maybe a size limit like 20MB?

**Comment by tech4GT** [comment]
@jywarren For now I used a common string but I'll update it to this following architecture
```js
{
// request body
   images: [
         {
          url: <URL>
          steps: <String> 
         },
         {
          url: <>
          steps: <>
         }
   ]
}
```

**Comment by tech4GT** [comment]
We can then expand each of the images with as much metadata as we want like coordinates and orientation.

**Comment by jywarren** [comment]
Oh but the images could be very large, we're only transmitting their URLs
on the initial request. I think they may be over 100mb once downloaded.

**Comment by tech4GT** [comment]
But it's gonna be that anyway, right? I mean if we transfer the image in part initially from the client side, then too the total data remains the same, I think the best option we have is to compress them in mapknitter and host those images on the urls. We can then create a step like decompress which gets the original image back.

**Comment by jywarren** [comment]
Ah I see. But the image may be remotely stored in aws or Google cloud, so
we should standardize on sending the URLs first. Later I agree we could
think of transmission speed optimization by compressing, but we might be
able to use on-the-fly gzip in nginx configuration, and do it at that level
instead of building it into the architecture. Sound good?

**Comment by tech4GT** [comment]
Yeah this sounds good to me!! :+1:

**Comment by tech4GT** [comment]
Hi @jywarren , I updated the api structure and now on using this url
`http://localhost:4000/api/v1/process/?images=[{"url":"./Images/Mario.png","sequence":"invert"},{"url":"./Images/Mario.png","sequence":"colormap"}]`

this is the result:
<img width="1440" alt="Screen Shot 2019-03-22 at 2 08 08 PM" src="https://user-images.githubusercontent.com/25617855/54810366-f381cf00-4cab-11e9-9fc9-bcc8364b10be.png">

**Comment by tech4GT** [comment]
I think this completes our part 1 majorly? Should I get started on the stitch module using https://github.com/publiclab/mapknitter-exporter/blob/main/lib/mapknitterExporter.rb#L252 as reference??

**Comment by jywarren** [comment]
OK, this is good; however, I wonder if we should give each incoming image
an id, and actually make it a step, in an associative array, so:


/api/v1/process/?steps={1:{"url":"./Images/Mario.png","sequence":"invert"},2:{"url":"./Images/Mario.png","sequence":"colormap"}}

Is this valid JSON?

But then we could add steps that run once other ones are done, but we need
to think how we'd represent this... could it be like:

/api/v1/process/?steps={1:{"url":"./Images/Mario.png","sequence":"invert"},2:{"url":"./Images/Mario.png","sequence":"colormap"}},3:{'after':
1,"sequence":"composite{image:$1},composite{image:$2}"}

What do you think? any suggestions?

**Comment by tech4GT** [comment]
Hmm, this sounds right! Maybe what we can do is have a common sequence that runs once others are done and then every image can have a sequence property which is specific to that image
```js
{
images: [
    {
       url: <>,
       sequence: <>
    }
],
sequence: <Common sequence>
}
```

**Comment by tech4GT** [comment]
Once we get this sorted we can move on to part 2 finally :smiley:

**Comment by jywarren** [comment]
I like separating steps from images because we can then use images more
than once, you know? And if we mark some steps as depending on others,
using their unique IDs, we can later optimize which can run in parallel.
What do you think?

**Comment by tech4GT** [comment]
Hmm, Sure! So should I update it according to the format you suggested earlier?
One thing I can't wrap my head around though, is that how would we use these IDs to generate the dependency info?

**Comment by jywarren** [comment]
I'm thinking,

1. we can have a trigger that runs at the start and whenever a step
completes,
2. then we check each next step with a conditional that checks if it's
prerequisites are complete, and starts them if so
3. when they complete, they also check remaining steps in order to see if
their prerequisites are complete

I'm brainstorming here, so maybe we need to think this through more before
implementing, but what do you think here?

**Comment by tech4GT** [comment]
Hello Jeff, sorry for the late reply! Okay this sounds good!!
- Every Image has an ID which is used for ordering and dependency resolution
- Inside the each image object we can have the properties, `pre` and `post` which will be specific to that particular image.
- We also have a `requisites` property on the image which is an array of IDs which need to be completed before this can be started.

**Comment by tech4GT** [comment]
This makes a lot of sense, I was meaning to ask that if we don't mind using some extra memory, I can implement a stripped down version of the `directed graph data structure`. We can then run a topological sorting algorithm on it which will take care of the dependencies. How does that sound?

**Comment by tech4GT** [comment]
So the API request will kind of look like
```js
{
images: [
   {
      <ID>
      <pre>
      <post>
      <sequence>
      <requisites>
   }
],
common_sequence: <String>
upload: <Boolean>
redirect: <Boolean>
}
```

**Comment by tech4GT** [comment]
cc @jywarren

**Comment by harshkhandeparkar** [comment]
* I had an idea, since upload times can be slow, how about creating multiple requests instead of one and each request can have a special ID.
* Each request can contain say *n* images or images of total size n MB. For each request with the same ID, an asynchronous (separate) thread can be opened with a maximum of let's say *t* threads. Each thread will process the img(s) it receives and put it in some kind of an output queue.
* Each output queue will have it's own ID(same as that of the request) and as the requests come in, images will be processed asynchronously.
* After processing each image and creating the output queue, the images can be sent back to the same ID asynchronously
###### OR
* Every request can have an index and the total number of indices can be put in every request. So if there are *x* requests, the first request will have an index **0**, the next one will have **1** and so on. The server can process the requests asynchronously and put them in the output queue in the order of their indices.
* The API can listen for more requests, if requests with some indices haven't arrived (with a timeout of course)
* Finally the whole output queue can be sent at once or asynchronously (to reduce download times as well).
* As old requests are processed, threads will be freed and new ones can be processed asynchronously.
* Again the maximum number of requests that can be sent with the same ID should also have a limit (i think).

**Comment by harshkhandeparkar** [comment]
cc @jywarren @tech4GT

**Comment by harshkhandeparkar** [comment]
One thing is that the frontend will also have to be developed to calculate the no of requests and no of images per request and to distribute those images accordingly.

**Comment by harshkhandeparkar** [comment]
* If the frontend isn't developed for this kind of behaviour, a single request with all the imgs can also be taken.
* So the server has to get the no of images `i` and the the number of requests `r`. If the number of requests is small, each request should be allowed a smaller no of images for the asynchronous processing to work efficiently. If there is a single request, no of imgs per request can be more as only a single thread will be utilised for it and no performance will be reduced.

### Idea 2
* If the inet speed of the user is fast and the user uploads a lot of images at once in the same request, the server can distribute the work itself between different threads and provide the full output at once (because inet speed is fast, that won't be a bottleneck).

**Comment by jywarren** [comment]
So one thing re: upload time is that we'll probably start by simply passing
a JSON post request to this app with URLs for each of the images, rather
than passing the full images in the request. Then the app can fetch them
when it wants. We might add ways to POST images up later, but this will be
simpler to start with.

> directed graph data structure

Whoa, so would this be able to figure out the optimal pairings of image
merges that could be done in the "tree" diagram I posted here?
https://github.com/publiclab/mapknitter/issues/298

If so, then GREAT! Although we again can start by just working linearly,
but this JSON structure for submitting requests can be easily adapted to
such a sorting system later if we use this "requisites" idea. Great!!!

Do you think it's better to just have a flat set of steps, instead of
pre/post for each image? I'm on the fence, but maybe we could plan to add
pre/post later since it is nice but adds a layer of complexity.

> the server can distribute the work itself between different threads and
provide the full output at once

Exactly, this is great Harsh - yes, this app should be able to send Google
Cloud function calls to do some tasks off-app and parallelize!

Thanks all!!!

**Comment by harshkhandeparkar** [comment]
@jywarren

> passing a JSON post request to this app with URLs 

Where would the URLs link to? Some other site like imgur? If yes, how will the user upload images to that site? Wouldn't that be a bottleneck?

**Comment by harshkhandeparkar** [comment]
Also is there a way to "stream" the images? I mean the rgba data values for each pixel can be sent in a stream for a max of a few imgs at once. The output can also be received in a stream.
This method can only be used for a few modules like contrast which perform the same func on each pixel(on use changePixel func).
Maybe the optimal method can be chosen by the user(or frontend) at runtime based on the requirements. The API should just support different options.

**Comment by harshkhandeparkar** [comment]
My concern is that process time(local processing time) should not be less than the upload time for multiple imgs so parallelizing upload, processing and download I think can be a good way to do this. I'm not sure though.

**Comment by harshkhandeparkar** [comment]
* Something like the user can create a secure connection with the server instead of just firing a request and waiting for a response.
* The input data can be streamed in case there is a lot of processing to be done. Instead of streaming bit by bit or byte by byte, we can stream image by image.
* Say 10 imgs are to be processed.
1. The first image is uploaded and is being processed.
2. In the mean while the 2nd image was uploaded and is being processed on a 2nd thread and the 1st img had a lot of complex steps like blur and edgeDetect.
3. The 3rd and 4th images were uploaded fast as they were small in size and were added to the queue.
4. The 5th img is uploading and it is a big one. 2nd img and 4th img finished processing as they had less steps.
5. 2nd image is being downloaded and 4th is in the download queue.
6. In the meanwhile, 5th image was uploaded and is being processed.
7. 6th and 7th images were again very small, were uploaded very fast, had less steps so were processed very fast.
8. 2nd, 4th and 1st were downloaded and 3rd image is being downloaded while 6th and 7th are in queue and 8th is being uploaded.
9. And so on....

**Comment by tech4GT** [comment]
@jywarren yes the sorting step should allow us to generate that structure which you have sketched on the other issue, and it would be really helpful if you can update this issue with the final request structure, I am not very clear about how we handle sequences for each image. :)

**Comment by jywarren** [comment]
So, any pre/post sequences would just be added to the general steps array for now; so there'd be just one list of steps to run. 

```js
{
images: [
   {
  id: 1,
  url: 'path'
   }
],
steps: [
  {
    id: 1,
    img: 1,
    steps: { sequencerSteps }
  },
  {
    id: 2,
    dependsOn: [1],
    img: 1,
    steps: { sequencerSteps }
  },
  {
    id: 3,
    dependsOn: [1,2],
    img: 2,
    steps: { sequencerSteps }
  }
]
}
```

Only question left i think is... can we access the output of an earlier step we depended on? 

Maybe there's a function like `getOutputOf(index)`

Or some way to specify `img` to be the output of an earlier step instead of an image?

**Comment by jywarren** [comment]
Wait, what if we collapsed images array into steps, so each step instead of referencing an image id, had an input attribute which is a URL?

For those steps which depend on an earlier step, we could say `input: "$1"`, meaning the output of step 1.

What do you think?

**Comment by tech4GT** [comment]
Hi @jywarren , Yeah this makes sense!!
This should work!! Now I can use the steps information to create a dependency graph of sorts and try to topologically sort it!!

**Comment by jywarren** [comment]
So, just a steps array, I guess...

```js
{
steps: [
  {
    id: 1,
    input: "https://example.com/image.png",
    steps: { sequencerSteps }
  },
  {
    id: 2,
    depends: [1], // only needed if other than input image
    input: "https://example.com/image.png",
    steps: { sequencerSteps }
  },
  {
    id: 3,
    depends: [1, 2],
    input: 2, // the output of step 2
    steps: { sequencerSteps }
  }
],
upload: false
}
```

Any downsides?

**Comment by tech4GT** [comment]
The idea of collapsing images into steps is super nice! Actually I was thinking kind of the same thing but the other way around, like collapsing steps into images but this is way better!!

**Comment by tech4GT** [comment]
Yes this should work!

**Comment by jywarren** [comment]
I just edited last comment, take a look! Awesome!

**Comment by tech4GT** [comment]
Well one downside is that this json will be a bit tedious, for instance if the same sequence needs to be applied to multiple images, we would have to specify it step by step, again and again.

**Comment by tech4GT** [comment]
But I think it's still the best of all options discussed so far!

**Comment by jywarren** [comment]
I think we still need depends because if the step does importImage, it may need more than the single input to be ready before starting. But maybe we could make it required only if it's asking for more than the single input?

**Comment by jywarren** [comment]
Ok let's implement this and if we learn more and come up with a new idea we can modify or add features. Thanks Varun!

**Comment by tech4GT** [comment]
@jywarren Updated your comment a little bit.

**Comment by tech4GT** [comment]
@jywarren Just pushed in the new code, have a look
Try this url
`http://localhost:4000/api/v1/process/?steps=[{%22id%22:1,%22input%22:%22./Images/Mario.png%22,%22steps%22:%22invert%22},{%22id%22:2,%22input%22:%22./Images/Mario.png%22,%22steps%22:%22brightness%22,%22depends%22:[1,3]},{%22id%22:3,%22input%22:1,%22steps%22:%22ndvi%22,%22depends%22:[1]},{%22id%22:4,%22input%22:3,%22steps%22:%22invert%22,%22depends%22:[2,3]}]`

this uses this dependency graph
1 -> 2
1 -> 3
3 -> 2
3 -> 4
2 -> 4

So the processing order is [1,3,2,4]

**Comment by jywarren** [comment]
Could you set this up in a Google cloud container with help from @icarito?
This is awesome!!!

**Comment by tech4GT** [comment]
@jywarren I think in the current state we can set this up as cloud function for demo purpose, should I do that??

**Comment by tech4GT** [comment]
@jywarren try `https://us-central1-public-lab.cloudfunctions.net/is-function/api/v1/process/?steps=[{%22id%22:1,%22input%22:%22https://publiclab.org/system/images/photos/000/006/671/original/canoe.jpg%22,%22steps%22:%22invert%22},{%22id%22:2,%22input%22:%22https://publiclab.org/system/images/photos/000/006/671/original/canoe.jpg%22,%22steps%22:%22brightness%22,%22depends%22:[1,3]},{%22id%22:3,%22input%22:1,%22steps%22:%22ndvi%22,%22depends%22:[1]},{%22id%22:4,%22input%22:3,%22steps%22:%22invert%22,%22depends%22:[2,3]}]`

**Comment by jywarren** [comment]
Wow it's amazing!!! So these are the outputs of each step? Awesome!

Ok, so perhaps time to circle back to webgl, no? What do you think of
attempting a parallel track of openCV.js? We'd be able to compare load
times for the headless-gl versus openCV approaches.

Fantastic work, Varun!

**Comment by tech4GT** [comment]
Yes they are the outputs of the steps.
Sure, since divy is trying out the headless browser track, I'll look into opencv.js
Do you think we can start the part 2 btw?

**Comment by jywarren** [comment]
Yes I think we're already in it even as there are a few leftovers from 1...
So, webgl distortion, compositing/canvas-size, and geotiff module? I think
those are the big remaining pieces of both parts, right?

Great work Varun!

**Comment by jywarren** [comment]
OK, so an incoming request would be based around the JSON format documented here:

https://github.com/publiclab/mapknitter-exporter-sinatra/pull/1#issuecomment-478173823

We'd need to develop a JS function to convert this into the format here:

https://github.com/publiclab/image-sequencer-app/issues/6#issuecomment-478205417

This could maybe live in this app for the time being, until we find a better home for it. I know it's mapknitter-specific and this app is broader, but we'll eventually integrate it into the Leaflet.DistortableImage library where we'll be able to export multiple selected images! https://github.com/publiclab/Leaflet.DistortableImage/issues/29

**Comment by tech4GT** [comment]
Sure @jywarren this sounds about right! But let's defer it until we get the actual workflow going so we know exactly what information to get.

**Comment by jywarren** [comment]
Shall we point at new `image-sequencer v3.3.0` and run on a new string...

`https://us-central1-public-lab.cloudfunctions.net/is-function/api/v1/process/?steps=[{"id":1,"input":"https://publiclab.org/i/31655.jpg","steps":"webgl-distort"}]`

`https://us-central1-public-lab.cloudfunctions.net/is-function/api/v1/process/?steps=[{"id":1,"input":"https://publiclab.org/i/31655.jpg","steps":"webgl-distort{nw:0,101|ne:1023,-51|se:1223,864|sw:100,764}"}]`

I'll try bumping `package.json` to get the new version.

**Comment by jywarren** [comment]
I also made a gist of the script: https://gist.github.com/jywarren/7721a38725affdbc4fad2598ff1be59e

**Comment by jywarren** [comment]
OK, getting very close here, it can't read the `|` syntax for properties...:

```js
Module 101|ne:1023 not found.
TypeError: Cannot read property 'description' of undefined 
at insertStep (/srv/node_modules/image-sequencer/src/InsertStep.js:23:50) 
at InsertStep (/srv/node_modules/image-sequencer/src/InsertStep.js:61:3) at AddStep (/srv/node_modules/image-sequencer/src/AddStep.js:3:33) 
at Object.addSteps (/srv/node_modules/image-sequencer/src/ImageSequencer.js:75:29) 
at eachStep (/srv/node_modules/image-sequencer/src/Strings.js:132:17) 
at Array.forEach (<anonymous>) 
at Object.importString (/srv/node_modules/image-sequencer/src/Strings.js:131:21) 
at Object.sequencerInstance.loadImages (/srv/index.js:64:27) 
at /srv/node_modules/image-sequencer/src/ImageSequencer.js:179:23 
at /srv/node_modules/image-sequencer/src/ui/LoadImage.js:67:7
```

**Comment by jywarren** [comment]
OK, removing the parameters, to just use the default `webgl-distort` settings, finally it tried to boot Chrome! 

`https://us-central1-public-lab.cloudfunctions.net/is-function/api/v1/process/?steps=[{"id":1,"input":"https://publiclab.org/i/31655.jpg","steps":"webgl-distort"}]`

But it did not succeed:

```
E  is-function n2cbn71yu75d Unhandled rejection is-function n2cbn71yu75d 
  undefined
E  is-function n2cbn71yu75d Error: Failed to launch chrome!
[0419/212419.930186:ERROR:zygote_host_impl_linux.cc(89)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.


TROUBLESHOOTING: https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md

    at onClose (/srv/node_modules/puppeteer/lib/Launcher.js:342:14)
    at Interface.helper.addEventListener (/srv/node_modules/puppeteer/lib/Launcher.js:331:50)
    at emitNone (events.js:111:20)
    at Interface.emit (events.js:208:7)
    at Interface.close (readline.js:368:8)
    at Socket.onend (readline.js:147:10)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:139:11) is-function n2cbn71yu75d 
  undefined
```

**Comment by jywarren** [comment]
https://github.com/karma-runner/karma-chrome-launcher/issues/158

**Comment by jywarren** [comment]
OK, also opened https://github.com/publiclab/image-sequencer/issues/1033 and https://github.com/publiclab/image-sequencer/issues/1032 which also need to be solved before this will run. 🎉

**Comment by jywarren** [comment]
Once we can confirm this is running, we'll need to make a converter function from the `input.json` format in https://github.com/publiclab/mapknitter-exporter-sinatra/pull/1#issuecomment-478173823 to this URL format shown here:

`https://us-central1-public-lab.cloudfunctions.net/is-function/api/v1/process/?steps=[{"id":1,"input":"https://publiclab.org/i/31655.jpg","steps":"webgl-distort{nw:0,101|ne:1023,-51|se:1223,864|sw:100,764}"}]`

**Comment by jywarren** [comment]
https://github.com/publiclab/image-sequencer/issues/1040

**Comment by jywarren** [comment]
OK, well -- with a small image, this works!!!

`https://us-central1-public-lab.cloudfunctions.net/is-function-edge/api/v1/process/?steps=[{%22id%22:1,%22input%22:%22https://i.publiclab.org/system/images/photos/000/000/354/thumb/Boots-ground-02.png%22,%22steps%22:%22webgl-distort%22}]`

Input: 

![](https://i.publiclab.org/system/images/photos/000/000/354/thumb/Boots-ground-02.png)

Output: 

![image](https://user-images.githubusercontent.com/24359/56988345-cbfd0c80-6b5d-11e9-9bb0-81e24e2279ab.png)

Here's a better test image:

![](https://user-images.githubusercontent.com/24359/56988552-55acda00-6b5e-11e9-8453-4df520e1b280.png)

****

OK! We had a JSON parse error but it seems to relate to the encoding of the punctuation marks; this URL will bypass it: 

https://us-central1-public-lab.cloudfunctions.net/is-function-edge/api/v1/process/?steps=[{"id":1,"input":"https://i.publiclab.org/i/31778.png","steps":"webgl-distort"},{"id":2,"input":"https://i.publiclab.org/i/31778.png","steps":"webgl-distort"}]

It's not completing... but it meets a timeout: `Function execution took 540002 ms, finished with status: 'timeout'`

> We either need to optimize a LOT, or to move to a non Cloud Functions approach where we can leave this open longer and/or increase power to the container we run it in. 

anyways, next step would be to try to pull in the output of one sequencer into the other, with a sequence like: `import-image{url:https://i.publiclab.org/i/31778.png},overlay{x:10|y:20}`

Actually, with the function code @tech4GT has written, we should be able to point at the output of the earlier step with a number as input, and then use overlay:

https://us-central1-public-lab.cloudfunctions.net/is-function-edge/api/v1/process/?steps=[{"id":1,"input":"https://i.publiclab.org/i/31778.png","steps":"webgl-distort"},{"id":2,"input":1,"steps":"import-image{url:https://i.publiclab.org/i/31778.png},overlay"}]

**Comment by jywarren** [comment]
Trying something that doesn't require Puppeteer, to confirm that we can do this `overlay` thing: https://us-central1-public-lab.cloudfunctions.net/is-function-edge/api/v1/process/?steps=[{%22id%22:1,%22input%22:%22https://publiclab.org/i/31655.jpg%22,%22steps%22:%22colormap%22},{%22id%22:2,%22input%22:1,%22steps%22:%22import-image{url:https://publiclab.org/i/31655.jpg},overlay%22}]

**Update:** OK! Got an error for the data-url here:

```
Bad image path { src: 'data:text/html; charset=utf-8;base64,PGh0bWw+PGJvZHk+WW91IGFyZSBiZWluZyA8YSBocmVmPSJodHRwczovL3B1YmxpY2xhYi5vcmcvc3lzdGVtL2ltYWdlcy9waG90b3MvMDAwLzAzMS82NTUvbGFyZ2UvREpJXzAwMjRfc21hbGwuanBnIj5yZWRpcmVjdGVkPC9hPi48L2JvZHk+PC9odG1sPg==', 
```

**Update:** haha, cool, that data-url wasn't an image -- it displayed:

`You are being redirected.`

**Comment by tech4GT** [comment]
@jywarren Is the parsing error specific to google cloud or are we facing it locally too?

**Comment by tech4GT** [comment]
Can you guide me to replicate the error locally?

**Comment by jywarren** [comment]
I found that I could get it to happen on Google Cloud, but ALSO using the
CLI interface locally.

**Comment by jywarren** [comment]
https://github.com/publiclab/image-sequencer/issues/1033 shows how to replicate it in the CLI mode - it gives the same error, so it seems likely it's the same parsing issue!

**Comment by jywarren** [comment]
I'm sorry - wait. The error is not the same. But @VibhorCodecianGupta has gone into some detail on the parsing in there, so there is more context available in https://github.com/publiclab/image-sequencer/issues/1033.

**Comment by jywarren** [comment]
So, i'd go by this comment to replicate it: https://github.com/publiclab/image-sequencer-app/issues/6#issuecomment-485018092

And I bet you could get this running locally: https://gist.github.com/jywarren/7721a38725affdbc4fad2598ff1be59e

**Comment by tech4GT** [comment]
@jywarren So to sum this up, the problem is with using `|` marks with our regular express framework, correct?
The query that is being passed to the sequencer gets messed up on using them as we did in the urls above, so this should happen locally too, right?
I tried running webgl-distort with options on the app running locally on my machine!
Great!! I got the error locally!!
On this now.

**Comment by tech4GT** [comment]
Okay I have isolated this error to our import string functions. Okay I'll make sure to write some tests for the string parsing code to prevent this in the future. :+1:

**Comment by tech4GT** [comment]
Also @jywarren I found out why the topological sort wasn't working! Actually that one is my fault since I didn't factor in that the graph might have nodes which are entirely disconnected. The library we are using does not support such nodes so all we need to do make that work is add those nodes in a loop. I'll make the necessary changes in a separate PR. :tada:

**Comment by jywarren** [comment]
Oh this is great to hear you reproduced it!

**Comment by jywarren** [comment]
Also this issue with the data url returned -- after these two fixes, would you mind taking a look? Thanks!

https://github.com/publiclab/image-sequencer-app/issues/6#issuecomment-488102548

**Comment by jywarren** [comment]
And Varun, if we get this version of multi-sequencer (from cloud functions) running in a Docker container, we can move it to a non Cloud Functions part of Google Cloud and get longer runtimes and more memory. I think @icarito can help us with this. So, let's aim to get all these bugs fixed and we'll ask @icarito for some help getting this booting in a docker container?

**Comment by tech4GT** [comment]
Sure! The project is already set up with docker, so it shouldn't be a problem adding the multi sequencer! 😄

**Comment by tech4GT** [comment]
Okay finally figured this out! The commas in the input for webgl-distort are messing with the commas separating steps!

**Comment by tech4GT** [comment]
@jywarren the problem is with the sequencer string not being URL safe. So how importString works is the inputs to every step, that is the stuff between `{...}` needs to be run through encodeURIComponent. So `webgl-distort{nw:0,101|ne:1023,-51|se:1223,868|sw:100,764}` will not work but this will `webgl-distort{nw:0%2C101|ne:1023%2C-51|se:1223%2C868|sw:100%2C764}`

**Comment by tech4GT** [comment]
I did it this way to make sure that complete functions can be passed in as arguments like in the blend step. :smile:

**Comment by tech4GT** [comment]
Okay also closing this up, since we have fixed the API structure and implemented parallel processing!
